File: packages/schema-domain/src/parent-owned-sync-export.ts
/* thin adapter over Rust-generated parent-owned sync export contracts */

import { type Infer, Schema, brandedNonEmptyStringSchema, withParser } from './effect';
import { countProductionProofValues } from './proof-shape';
import {
  FamilyReferenceSchema,
  ParentActionReferenceSchema,
  ParentDeviceReferenceSchema,
} from './family-references';
import type {
  ChildProfileId as ChildProfileIdType,
  FamilyId as FamilyIdType,
  ParentActionReferenceId as ParentActionReferenceIdType,
  ParentActorId as ParentActorIdType,
  ParentDeviceId as ParentDeviceIdType,
  ParentDeviceLabel as ParentDeviceLabelType,
  ParentEvidenceReferenceId as ParentEvidenceReferenceIdType,
  ParentPolicyVersion as ParentPolicyVersionType,
  ParentTimestamp as ParentTimestampType,
} from './family-reference-primitives';
import { ParentContractSchemaVersionSchema, ParentTimestampSchema } from './family-reference-primitives';
import {
  GeneratedParentOwnedSyncDeleteVisibilityStates,
  GeneratedParentOwnedSyncDisconnectVisibilityStates,
  GeneratedParentOwnedSyncExportContractProof,
  GeneratedParentOwnedSyncExportDataClasses,
  GeneratedParentOwnedSyncExportDestinationOwnerships,
  GeneratedParentOwnedSyncExportEncryptionStates,
  GeneratedParentOwnedSyncExportFormats,
  GeneratedParentOwnedSyncExportKnownGaps,
  GeneratedParentOwnedSyncExportNonClaims,
  type GeneratedParentOwnedSyncExportManifestItem,
  GeneratedParentOwnedSyncManifestIntegrityStates,
  GeneratedParentOwnedSyncProviderModes,
  GeneratedParentOwnedSyncProviderStatuses,
  GeneratedParentOwnedSyncStates,
  GeneratedParentOwnedSyncTombstonePropagationStates,
  ParentOwnedSyncExportContractRuntime,
  type GeneratedParentOwnedSyncExportContractProof as GeneratedParentOwnedSyncExportContractProofShape,
  type GeneratedParentOwnedSyncProviderStatusRow,
  type GeneratedParentOwnedSyncStateRow,
  type GeneratedParentOwnedSyncTombstoneRow,
} from './generated/parent-owned-sync-export-contracts';
import {
  syncExportContractProofIsHonestGenerated,
  syncExportCoversRequiredDataClassesGenerated,
  syncExportManifestItemIsHonestGenerated,
  syncExportProviderStatusRowIsHonestGenerated,
  syncExportSyncStateRowIsHonestGenerated,
  syncExportTombstoneRowIsHonestGenerated,
} from './generated/parent-owned-sync-export-contract-rules';

export const ParentOwnedSyncExportSchemaVersionSchema = withParser(
  Schema.Literal(ParentOwnedSyncExportContractRuntime.SchemaVersion)
);

export const ParentOwnedSyncExportDataClassSchema = withParser(
  Schema.Literal(...GeneratedParentOwnedSyncExportDataClasses)
);
export const ParentOwnedSyncExportFormatSchema = withParser(Schema.Literal(...GeneratedParentOwnedSyncExportFormats));
export const ParentOwnedSyncExportDestinationOwnershipSchema = withParser(
  Schema.Literal(...GeneratedParentOwnedSyncExportDestinationOwnerships)
);
export const ParentOwnedSyncExportEncryptionStateSchema = withParser(
  Schema.Literal(...GeneratedParentOwnedSyncExportEncryptionStates)
);
export const ParentOwnedSyncExportProviderModeSchema = withParser(
  Schema.Literal(...GeneratedParentOwnedSyncProviderModes)
);
export const ParentOwnedSyncExportProviderStatusSchema = withParser(
  Schema.Literal(...GeneratedParentOwnedSyncProviderStatuses)
);
export const ParentOwnedSyncExportManifestIntegrityStateSchema = withParser(
  Schema.Literal(...GeneratedParentOwnedSyncManifestIntegrityStates)
);
export const ParentOwnedSyncExportSyncStateSchema = withParser(Schema.Literal(...GeneratedParentOwnedSyncStates));
export const ParentOwnedSyncExportTombstonePropagationStateSchema = withParser(
  Schema.Literal(...GeneratedParentOwnedSyncTombstonePropagationStates)
);
export const ParentOwnedSyncExportDisconnectVisibilityStateSchema = withParser(
  Schema.Literal(...GeneratedParentOwnedSyncDisconnectVisibilityStates)
);
export const ParentOwnedSyncExportDeleteVisibilityStateSchema = withParser(
  Schema.Literal(...GeneratedParentOwnedSyncDeleteVisibilityStates)
);
export const ParentOwnedSyncExportNonClaimSchema = withParser(
  Schema.Literal(...GeneratedParentOwnedSyncExportNonClaims)
);

const ParentOwnedSyncManifestIdSchema = brandedNonEmptyStringSchema('ParentOwnedSyncManifestId');
const ParentOwnedSyncItemIdSchema = brandedNonEmptyStringSchema('ParentOwnedSyncItemId');
const ParentOwnedSyncVersionLabelSchema = brandedNonEmptyStringSchema('ParentOwnedSyncVersionLabel');
const ParentOwnedSyncPolicyRefSchema = brandedNonEmptyStringSchema('ParentOwnedSyncPolicyRef');
const ParentOwnedSyncProviderIdSchema = brandedNonEmptyStringSchema('ParentOwnedSyncProviderId');
const ParentOwnedSyncProviderRefSchema = brandedNonEmptyStringSchema('ParentOwnedSyncProviderRef');
const ParentOwnedSyncStatusRefSchema = brandedNonEmptyStringSchema('ParentOwnedSyncStatusRef');
const ParentOwnedSyncCursorRefSchema = brandedNonEmptyStringSchema('ParentOwnedSyncCursorRef');
const ParentOwnedSyncBatchRefSchema = brandedNonEmptyStringSchema('ParentOwnedSyncBatchRef');
const ParentOwnedSyncConflictRefSchema = brandedNonEmptyStringSchema('ParentOwnedSyncConflictRef');
const ParentOwnedSyncChecksumRefSchema = brandedNonEmptyStringSchema('ParentOwnedSyncChecksumRef');
const ParentOwnedSyncSignatureRefSchema = brandedNonEmptyStringSchema('ParentOwnedSyncSignatureRef');
const ParentOwnedSyncTombstoneRefSchema = brandedNonEmptyStringSchema('ParentOwnedSyncTombstoneRef');
const ParentOwnedSyncDeleteRequestRefSchema = brandedNonEmptyStringSchema('ParentOwnedSyncDeleteRequestRef');

export type FamilyId = FamilyIdType;
export type ChildProfileId = ChildProfileIdType;
export type ParentDeviceId = ParentDeviceIdType;
export type ParentDeviceLabel = ParentDeviceLabelType;
export type ParentActorId = ParentActorIdType;
export type ParentPolicyVersion = ParentPolicyVersionType;
export type ParentEvidenceReferenceId = ParentEvidenceReferenceIdType;
export type ParentActionReferenceId = ParentActionReferenceIdType;
export type ParentTimestamp = ParentTimestampType;
export type ParentOwnedSyncManifestId = typeof ParentOwnedSyncManifestIdSchema.Type;
export type ParentOwnedSyncItemId = typeof ParentOwnedSyncItemIdSchema.Type;
export type ParentOwnedSyncVersionLabel = typeof ParentOwnedSyncVersionLabelSchema.Type;
export type ParentOwnedSyncPolicyRef = typeof ParentOwnedSyncPolicyRefSchema.Type;
export type ParentOwnedSyncProviderId = typeof ParentOwnedSyncProviderIdSchema.Type;
export type ParentOwnedSyncProviderRef = typeof ParentOwnedSyncProviderRefSchema.Type;
export type ParentOwnedSyncStatusRef = typeof ParentOwnedSyncStatusRefSchema.Type;
export type ParentOwnedSyncCursorRef = typeof ParentOwnedSyncCursorRefSchema.Type;
export type ParentOwnedSyncBatchRef = typeof ParentOwnedSyncBatchRefSchema.Type;
export type ParentOwnedSyncConflictRef = typeof ParentOwnedSyncConflictRefSchema.Type;
export type ParentOwnedSyncChecksumRef = typeof ParentOwnedSyncChecksumRefSchema.Type;
export type ParentOwnedSyncSignatureRef = typeof ParentOwnedSyncSignatureRefSchema.Type;
export type ParentOwnedSyncTombstoneRef = typeof ParentOwnedSyncTombstoneRefSchema.Type;
export type ParentOwnedSyncDeleteRequestRef = typeof ParentOwnedSyncDeleteRequestRefSchema.Type;

const ParentOwnedSyncExportEncryptionMetadataBaseSchema = Schema.Struct({
  encryptionState: ParentOwnedSyncExportEncryptionStateSchema,
  encryptedBeforeUpload: Schema.Boolean,
  keyOwner: ParentOwnedSyncExportDestinationOwnershipSchema,
  proofRequirementRef: ParentOwnedSyncPolicyRefSchema,
});

export const ParentOwnedSyncExportEncryptionMetadataSchema = withParser(
  ParentOwnedSyncExportEncryptionMetadataBaseSchema
);

const ParentOwnedSyncExportManifestItemBaseSchema = Schema.Struct({
  itemId: ParentOwnedSyncItemIdSchema,
  dataClass: ParentOwnedSyncExportDataClassSchema,
  exportFormat: ParentOwnedSyncExportFormatSchema,
  destinationOwnership: ParentOwnedSyncExportDestinationOwnershipSchema,
  schemaVersionLabel: ParentOwnedSyncVersionLabelSchema,
  encryption: ParentOwnedSyncExportEncryptionMetadataSchema,
  parentActionRequired: Schema.Boolean,
  rawChildEvidenceUploadedByDefault: Schema.Boolean,
  ocentraHostedFamilyDataStored: Schema.Boolean,
  claimSafe: Schema.Boolean,
});

export const ParentOwnedSyncExportManifestItemSchema = withParser(
  ParentOwnedSyncExportManifestItemBaseSchema.pipe(
    Schema.filter(
      (item) =>
        syncExportManifestItemIsHonestGenerated(item as GeneratedParentOwnedSyncExportManifestItem) ||
        'Expected manifest items to stay encrypted-before-upload or explicitly human-readable, claim-safe, and free of hosted evidence storage'
    )
  )
);

const ParentOwnedSyncExportManifestBaseSchema = Schema.Struct({
  schemaVersion: ParentOwnedSyncExportSchemaVersionSchema,
  manifestId: ParentOwnedSyncManifestIdSchema,
  family: FamilyReferenceSchema,
  device: ParentDeviceReferenceSchema,
  parentAction: ParentActionReferenceSchema,
  productVersion: ParentOwnedSyncVersionLabelSchema,
  manifestVersion: ParentOwnedSyncVersionLabelSchema,
  generatedAt: ParentTimestampSchema,
  items: Schema.Array(ParentOwnedSyncExportManifestItemSchema),
});

export const ParentOwnedSyncExportManifestSchema = withParser(
  ParentOwnedSyncExportManifestBaseSchema.pipe(
    Schema.filter(
      (manifest) =>
        syncExportCoversRequiredDataClassesGenerated(
          manifest.items as ReadonlyArray<GeneratedParentOwnedSyncExportManifestItem>
        ) ||
        'Expected the parent-owned sync manifest to cover every required data class'
    )
  )
);

const ParentOwnedSyncExportProviderStatusRowBaseSchema = Schema.Struct({
  providerId: ParentOwnedSyncProviderIdSchema,
  providerMode: ParentOwnedSyncExportProviderModeSchema,
  providerStatus: ParentOwnedSyncExportProviderStatusSchema,
  destinationOwnership: ParentOwnedSyncExportDestinationOwnershipSchema,
  accountRef: Schema.Union(ParentOwnedSyncProviderRefSchema, Schema.Null),
  folderRef: Schema.Union(ParentOwnedSyncProviderRefSchema, Schema.Null),
  statusRef: ParentOwnedSyncStatusRefSchema,
  revocationRef: Schema.Union(ParentOwnedSyncProviderRefSchema, Schema.Null),
  disconnectVisibilityState: ParentOwnedSyncExportDisconnectVisibilityStateSchema,
  deleteVisibilityState: ParentOwnedSyncExportDeleteVisibilityStateSchema,
  lastCheckedAt: ParentTimestampSchema,
  oauthRuntimeClaimed: Schema.Boolean,
  uploadRuntimeClaimed: Schema.Boolean,
  deleteRuntimeClaimed: Schema.Boolean,
  ocentraHostedFamilyDataStored: Schema.Boolean,
  claimSafe: Schema.Boolean,
});

export const ParentOwnedSyncExportProviderStatusRowSchema = withParser(
  ParentOwnedSyncExportProviderStatusRowBaseSchema.pipe(
    Schema.filter(
      (row) =>
        syncExportProviderStatusRowIsHonestGenerated(row as GeneratedParentOwnedSyncProviderStatusRow) ||
        'Expected provider status rows to keep mode split, claim-safe status, and delete or disconnect visibility explicit without runtime overclaim'
    )
  )
);

const ParentOwnedSyncExportSyncStateRowBaseSchema = Schema.Struct({
  syncState: ParentOwnedSyncExportSyncStateSchema,
  providerStatusRef: ParentOwnedSyncStatusRefSchema,
  cursorRef: Schema.Union(ParentOwnedSyncCursorRefSchema, Schema.Null),
  batchRef: Schema.Union(ParentOwnedSyncBatchRefSchema, Schema.Null),
  manifestIntegrityState: ParentOwnedSyncExportManifestIntegrityStateSchema,
  manifestChecksumRef: Schema.Union(ParentOwnedSyncChecksumRefSchema, Schema.Null),
  manifestSignatureRef: Schema.Union(ParentOwnedSyncSignatureRefSchema, Schema.Null),
  lastSuccessfulSyncAt: Schema.Union(ParentTimestampSchema, Schema.Null),
  conflictRef: Schema.Union(ParentOwnedSyncConflictRefSchema, Schema.Null),
  retryQueueRef: Schema.Union(ParentOwnedSyncPolicyRefSchema, Schema.Null),
  parentActionRequired: Schema.Boolean,
  claimSafe: Schema.Boolean,
});

export const ParentOwnedSyncExportSyncStateRowSchema = withParser(
  ParentOwnedSyncExportSyncStateRowBaseSchema.pipe(
    Schema.filter(
      (row) =>
        syncExportSyncStateRowIsHonestGenerated(row as GeneratedParentOwnedSyncStateRow) ||
        'Expected sync states to keep checksum, signature, conflict, retry, corruption, and manual-required boundaries explicit'
    )
  )
);

const ParentOwnedSyncExportTombstoneRowBaseSchema = Schema.Struct({
  tombstoneRef: ParentOwnedSyncTombstoneRefSchema,
  dataClass: ParentOwnedSyncExportDataClassSchema,
  propagationState: ParentOwnedSyncExportTombstonePropagationStateSchema,
  deleteRequestRef: Schema.Union(ParentOwnedSyncDeleteRequestRefSchema, Schema.Null),
  providerStatusRef: ParentOwnedSyncStatusRefSchema,
  lastPropagatedAt: Schema.Union(ParentTimestampSchema, Schema.Null),
  blockedReasonRef: Schema.Union(ParentOwnedSyncPolicyRefSchema, Schema.Null),
  claimSafe: Schema.Boolean,
});

export const ParentOwnedSyncExportTombstoneRowSchema = withParser(
  ParentOwnedSyncExportTombstoneRowBaseSchema.pipe(
    Schema.filter(
      (row) =>
        syncExportTombstoneRowIsHonestGenerated(row as GeneratedParentOwnedSyncTombstoneRow) ||
        'Expected tombstone propagation to stay explicit and separate from sync success or provider status'
    )
  )
);

const ParentOwnedSyncExportContractProofBaseSchema = Schema.Struct({
  schemaVersion: ParentOwnedSyncExportSchemaVersionSchema,
  contractVersion: ParentContractSchemaVersionSchema,
  manifest: ParentOwnedSyncExportManifestSchema,
  providerStatuses: Schema.Array(ParentOwnedSyncExportProviderStatusRowSchema),
  syncStates: Schema.Array(ParentOwnedSyncExportSyncStateRowSchema),
  tombstones: Schema.Array(ParentOwnedSyncExportTombstoneRowSchema),
  nonClaims: Schema.Array(ParentOwnedSyncExportNonClaimSchema),
  transferRuntimeClaimed: Schema.Boolean,
  connectorOAuthClaimed: Schema.Boolean,
  uploadRuntimeClaimed: Schema.Boolean,
  deleteRuntimeClaimed: Schema.Boolean,
  ocentraHostedChildEvidenceStored: Schema.Boolean,
  updatedAt: ParentTimestampSchema,
});

export const ParentOwnedSyncExportContractProofSchema = withParser(
  ParentOwnedSyncExportContractProofBaseSchema.pipe(
    Schema.filter(
      (proof) =>
        syncExportContractProofIsHonestGenerated(proof as GeneratedParentOwnedSyncExportContractProofShape) ||
        'Expected parent-owned sync proof to keep provider, sync, manifest integrity, tombstone, and non-claim boundaries honest'
    )
  )
);

export type ParentOwnedSyncExportDataClass = Infer<typeof ParentOwnedSyncExportDataClassSchema>;
export type ParentOwnedSyncExportFormat = Infer<typeof ParentOwnedSyncExportFormatSchema>;
export type ParentOwnedSyncExportDestinationOwnership = Infer<
  typeof ParentOwnedSyncExportDestinationOwnershipSchema
>;
export type ParentOwnedSyncExportProviderMode = Infer<typeof ParentOwnedSyncExportProviderModeSchema>;
export type ParentOwnedSyncExportProviderStatus = Infer<typeof ParentOwnedSyncExportProviderStatusSchema>;
export type ParentOwnedSyncExportSyncState = Infer<typeof ParentOwnedSyncExportSyncStateSchema>;
export type ParentOwnedSyncExportTombstonePropagationState = Infer<
  typeof ParentOwnedSyncExportTombstonePropagationStateSchema
>;
export type ParentOwnedSyncExportNonClaim = Infer<typeof ParentOwnedSyncExportNonClaimSchema>;
export type ParentOwnedSyncExportManifestItem = GeneratedParentOwnedSyncExportManifestItem;
export type ParentOwnedSyncExportProviderStatusRow = GeneratedParentOwnedSyncProviderStatusRow;
export type ParentOwnedSyncExportSyncStateRow = GeneratedParentOwnedSyncStateRow;
export type ParentOwnedSyncExportTombstoneRow = GeneratedParentOwnedSyncTombstoneRow;
export type ParentOwnedSyncExportContractProofCandidate = GeneratedParentOwnedSyncExportContractProofShape;
export type ParentOwnedSyncExportContractProof = GeneratedParentOwnedSyncExportContractProofShape;

export const RequiredParentOwnedSyncExportProviderModes = [...GeneratedParentOwnedSyncProviderModes] as const;
export const RequiredParentOwnedSyncExportProviderStatuses = [...GeneratedParentOwnedSyncProviderStatuses] as const;
export const RequiredParentOwnedSyncExportSyncStates = [...GeneratedParentOwnedSyncStates] as const;
export const RequiredParentOwnedSyncExportTombstoneStates = [
  ...GeneratedParentOwnedSyncTombstonePropagationStates,
] as const;
export const ParentOwnedSyncExportKnownGaps = [...GeneratedParentOwnedSyncExportKnownGaps] as const;

export const ParentOwnedSyncExportContractProofReadModel = ParentOwnedSyncExportContractProofSchema.parse(
  GeneratedParentOwnedSyncExportContractProof
);

export function summarizeParentOwnedSyncExportDataClasses(
  items: ReadonlyArray<ParentOwnedSyncExportManifestItem>
): Record<ParentOwnedSyncExportDataClass, number> {
  return countProductionProofValues(items.map((item) => item.dataClass), GeneratedParentOwnedSyncExportDataClasses);
}

export function summarizeParentOwnedSyncExportProviderModes(
  rows: ReadonlyArray<ParentOwnedSyncExportProviderStatusRow>
): Record<ParentOwnedSyncExportProviderMode, number> {
  return countProductionProofValues(rows.map((row) => row.providerMode), GeneratedParentOwnedSyncProviderModes);
}

export function summarizeParentOwnedSyncExportProviderStatuses(
  rows: ReadonlyArray<ParentOwnedSyncExportProviderStatusRow>
): Record<ParentOwnedSyncExportProviderStatus, number> {
  return countProductionProofValues(rows.map((row) => row.providerStatus), GeneratedParentOwnedSyncProviderStatuses);
}

export function summarizeParentOwnedSyncExportSyncStates(
  rows: ReadonlyArray<ParentOwnedSyncExportSyncStateRow>
): Record<ParentOwnedSyncExportSyncState, number> {
  return countProductionProofValues(rows.map((row) => row.syncState), GeneratedParentOwnedSyncStates);
}

export function summarizeParentOwnedSyncExportTombstoneStates(
  rows: ReadonlyArray<ParentOwnedSyncExportTombstoneRow>
): Record<ParentOwnedSyncExportTombstonePropagationState, number> {
  return countProductionProofValues(
    rows.map((row) => row.propagationState),
    GeneratedParentOwnedSyncTombstonePropagationStates
  );
}

File: crates/storage-custody-core/tests/unit/parent_owned_sync_export.rs
use crate::support::StorageCustodyTestValueExt;

use ocentra_schema::parent_owned_sync_export as contracts;
use ocentra_storage_custody_core::parent_owned_sync_export::{
    build_parent_owned_sync_export_proof, derive_parent_owned_sync_provider_status_row,
    derive_parent_owned_sync_state_row, derive_parent_owned_sync_tombstone_row,
    ParentOwnedSyncExportDerivationError, ParentOwnedSyncProviderStatusInput,
    ParentOwnedSyncStateInput, ParentOwnedSyncTombstoneInput,
};

#[test]
fn provider_modes_and_claim_safe_statuses_stay_explicit() {
    let proof = contracts::sample_parent_owned_sync_export_contract_proof();

    assert_eq!(proof.provider_statuses.len(), 11);
    assert!(proof
        .provider_statuses
        .iter()
        .all(|row| row.claim_safe && !row.oauth_runtime_claimed && !row.upload_runtime_claimed));
    assert!(
        proof
            .provider_statuses
            .iter()
            .any(|row| row.provider_status
                == contracts::ParentOwnedSyncProviderStatus::ManualRequired)
    );
    assert!(proof.provider_statuses.iter().any(|row| {
        row.disconnect_visibility_state
            == contracts::ParentOwnedSyncDisconnectVisibilityState::DisconnectVisible
    }));
}

#[test]
fn ready_and_revoked_provider_rows_require_their_specific_refs() {
    let timestamp = contracts::ParentTimestamp::parse("2026-06-28T18:40:00.000Z").assume_ok();
    let ready_missing_refs =
        derive_parent_owned_sync_provider_status_row(ParentOwnedSyncProviderStatusInput {
            provider_id: contracts::ParentOwnedSyncProviderId::parse("provider-ready").assume_ok(),
            provider_mode: contracts::ParentOwnedSyncProviderMode::GoogleDriveAppdata,
            provider_status: contracts::ParentOwnedSyncProviderStatus::Ready,
            destination_ownership:
                contracts::ParentOwnedSyncExportDestinationOwnership::ParentOwnedExternalStorage,
            account_ref: None,
            folder_ref: None,
            status_ref: contracts::ParentOwnedSyncStatusRef::parse("provider-status-ready")
                .assume_ok(),
            revocation_ref: None,
            disconnect_visibility_state:
                contracts::ParentOwnedSyncDisconnectVisibilityState::NotDisconnected,
            delete_visibility_state: contracts::ParentOwnedSyncDeleteVisibilityState::NotRequested,
            last_checked_at: timestamp.clone(),
        });
    assert_eq!(
        ready_missing_refs,
        Err(ParentOwnedSyncExportDerivationError::ReadyProviderMissingLocationRefs)
    );

    let revoked_missing_ref =
        derive_parent_owned_sync_provider_status_row(ParentOwnedSyncProviderStatusInput {
            provider_id: contracts::ParentOwnedSyncProviderId::parse("provider-revoked")
                .assume_ok(),
            provider_mode: contracts::ParentOwnedSyncProviderMode::OnedriveApproot,
            provider_status: contracts::ParentOwnedSyncProviderStatus::Revoked,
            destination_ownership:
                contracts::ParentOwnedSyncExportDestinationOwnership::ParentOwnedExternalStorage,
            account_ref: contracts::ParentOwnedSyncProviderRef::parse("account-revoked"),
            folder_ref: contracts::ParentOwnedSyncProviderRef::parse("folder-revoked"),
            status_ref: contracts::ParentOwnedSyncStatusRef::parse("provider-status-revoked")
                .assume_ok(),
            revocation_ref: None,
            disconnect_visibility_state:
                contracts::ParentOwnedSyncDisconnectVisibilityState::NotDisconnected,
            delete_visibility_state: contracts::ParentOwnedSyncDeleteVisibilityState::NotRequested,
            last_checked_at: timestamp,
        });
    assert_eq!(
        revoked_missing_ref,
        Err(ParentOwnedSyncExportDerivationError::RevokedProviderMissingRevocationRef)
    );
}

#[test]
fn sync_states_require_manifest_integrity_and_retry_evidence_honestly() {
    let synced_missing_signature = derive_parent_owned_sync_state_row(ParentOwnedSyncStateInput {
        sync_state: contracts::ParentOwnedSyncState::Synced,
        provider_status_ref: contracts::ParentOwnedSyncStatusRef::parse("provider-status-ready")
            .assume_ok(),
        cursor_ref: contracts::ParentOwnedSyncCursorRef::parse("cursor-synced"),
        batch_ref: contracts::ParentOwnedSyncBatchRef::parse("batch-synced"),
        manifest_integrity_state: contracts::ParentOwnedSyncManifestIntegrityState::Verified,
        manifest_checksum_ref: contracts::ParentOwnedSyncChecksumRef::parse("checksum-synced"),
        manifest_signature_ref: None,
        last_successful_sync_at: contracts::ParentTimestamp::parse("2026-06-28T18:40:00.000Z"),
        conflict_ref: None,
        retry_queue_ref: None,
        parent_action_required: false,
    });
    assert_eq!(
        synced_missing_signature,
        Err(
            ParentOwnedSyncExportDerivationError::SuccessfulSyncRequiresCursorBatchChecksumAndSignature
        )
    );

    let corrupt_synced = derive_parent_owned_sync_state_row(ParentOwnedSyncStateInput {
        sync_state: contracts::ParentOwnedSyncState::Synced,
        provider_status_ref: contracts::ParentOwnedSyncStatusRef::parse("provider-status-manual")
            .assume_ok(),
        cursor_ref: contracts::ParentOwnedSyncCursorRef::parse("cursor-corrupt"),
        batch_ref: contracts::ParentOwnedSyncBatchRef::parse("batch-corrupt"),
        manifest_integrity_state: contracts::ParentOwnedSyncManifestIntegrityState::Corrupt,
        manifest_checksum_ref: contracts::ParentOwnedSyncChecksumRef::parse("checksum-corrupt"),
        manifest_signature_ref: contracts::ParentOwnedSyncSignatureRef::parse("signature-corrupt"),
        last_successful_sync_at: contracts::ParentTimestamp::parse("2026-06-28T18:40:00.000Z"),
        conflict_ref: None,
        retry_queue_ref: None,
        parent_action_required: true,
    });
    assert_eq!(
        corrupt_synced,
        Err(ParentOwnedSyncExportDerivationError::CorruptManifestCannotClaimSynced)
    );

    let conflict_missing_ref = derive_parent_owned_sync_state_row(ParentOwnedSyncStateInput {
        sync_state: contracts::ParentOwnedSyncState::Conflict,
        provider_status_ref: contracts::ParentOwnedSyncStatusRef::parse("provider-status-conflict")
            .assume_ok(),
        cursor_ref: contracts::ParentOwnedSyncCursorRef::parse("cursor-conflict"),
        batch_ref: contracts::ParentOwnedSyncBatchRef::parse("batch-conflict"),
        manifest_integrity_state: contracts::ParentOwnedSyncManifestIntegrityState::Verified,
        manifest_checksum_ref: contracts::ParentOwnedSyncChecksumRef::parse("checksum-conflict"),
        manifest_signature_ref: contracts::ParentOwnedSyncSignatureRef::parse("signature-conflict"),
        last_successful_sync_at: None,
        conflict_ref: None,
        retry_queue_ref: contracts::ParentOwnedSyncPolicyRef::parse("retry-conflict"),
        parent_action_required: true,
    });
    assert_eq!(
        conflict_missing_ref,
        Err(ParentOwnedSyncExportDerivationError::ConflictStateRequiresConflictRef)
    );
}

#[test]
fn tombstone_propagation_stays_separate_from_sync_success() {
    let propagated = derive_parent_owned_sync_tombstone_row(ParentOwnedSyncTombstoneInput {
        tombstone_ref: contracts::ParentOwnedSyncTombstoneRef::parse("tombstone-propagated")
            .assume_ok(),
        data_class: contracts::ParentOwnedSyncExportDataClass::GeneratedSummary,
        propagation_state: contracts::ParentOwnedSyncTombstonePropagationState::Propagated,
        delete_request_ref: contracts::ParentOwnedSyncDeleteRequestRef::parse("delete-summary"),
        provider_status_ref: contracts::ParentOwnedSyncStatusRef::parse("provider-status-ready")
            .assume_ok(),
        last_propagated_at: contracts::ParentTimestamp::parse("2026-06-28T18:40:00.000Z"),
        blocked_reason_ref: None,
    })
    .assume_ok();
    assert_eq!(
        propagated.propagation_state,
        contracts::ParentOwnedSyncTombstonePropagationState::Propagated
    );

    let blocked_missing_reason =
        derive_parent_owned_sync_tombstone_row(ParentOwnedSyncTombstoneInput {
            tombstone_ref: contracts::ParentOwnedSyncTombstoneRef::parse("tombstone-blocked")
                .assume_ok(),
            data_class: contracts::ParentOwnedSyncExportDataClass::NotificationHistory,
            propagation_state: contracts::ParentOwnedSyncTombstonePropagationState::Blocked,
            delete_request_ref: contracts::ParentOwnedSyncDeleteRequestRef::parse(
                "delete-notification",
            ),
            provider_status_ref: contracts::ParentOwnedSyncStatusRef::parse(
                "provider-status-folder",
            )
            .assume_ok(),
            last_propagated_at: None,
            blocked_reason_ref: None,
        });
    assert_eq!(
        blocked_missing_reason,
        Err(ParentOwnedSyncExportDerivationError::TombstoneBlockedRequiresReason)
    );
}

#[test]
fn proof_builder_keeps_non_claims_and_no_default_ocentra_custody_truth() {
    let sample = contracts::sample_parent_owned_sync_export_contract_proof();
    let built = build_parent_owned_sync_export_proof(
        &sample.manifest,
        sample
            .provider_statuses
            .iter()
            .cloned()
            .map(|row| ParentOwnedSyncProviderStatusInput {
                provider_id: row.provider_id,
                provider_mode: row.provider_mode,
                provider_status: row.provider_status,
                destination_ownership: row.destination_ownership,
                account_ref: row.account_ref,
                folder_ref: row.folder_ref,
                status_ref: row.status_ref,
                revocation_ref: row.revocation_ref,
                disconnect_visibility_state: row.disconnect_visibility_state,
                delete_visibility_state: row.delete_visibility_state,
                last_checked_at: row.last_checked_at,
            })
            .collect(),
        sample
            .sync_states
            .iter()
            .cloned()
            .map(|row| ParentOwnedSyncStateInput {
                sync_state: row.sync_state,
                provider_status_ref: row.provider_status_ref,
                cursor_ref: row.cursor_ref,
                batch_ref: row.batch_ref,
                manifest_integrity_state: row.manifest_integrity_state,
                manifest_checksum_ref: row.manifest_checksum_ref,
                manifest_signature_ref: row.manifest_signature_ref,
                last_successful_sync_at: row.last_successful_sync_at,
                conflict_ref: row.conflict_ref,
                retry_queue_ref: row.retry_queue_ref,
                parent_action_required: row.parent_action_required,
            })
            .collect(),
        sample
            .tombstones
            .iter()
            .cloned()
            .map(|row| ParentOwnedSyncTombstoneInput {
                tombstone_ref: row.tombstone_ref,
                data_class: row.data_class,
                propagation_state: row.propagation_state,
                delete_request_ref: row.delete_request_ref,
                provider_status_ref: row.provider_status_ref,
                last_propagated_at: row.last_propagated_at,
                blocked_reason_ref: row.blocked_reason_ref,
            })
            .collect(),
        contracts::ParentTimestamp::parse("2026-06-28T18:50:00.000Z").assume_ok(),
    )
    .assume_ok();

    assert_eq!(
        built.non_claims,
        contracts::required_parent_owned_sync_export_non_claims()
    );
    assert!(!built.transfer_runtime_claimed);
    assert!(!built.connector_o_auth_claimed);
    assert!(!built.upload_runtime_claimed);
    assert!(!built.delete_runtime_claimed);
    assert!(!built.ocentra_hosted_child_evidence_stored);
    assert_eq!(built.provider_statuses.len(), 11);
    assert_eq!(built.sync_states.len(), 8);
    assert_eq!(built.tombstones.len(), 5);
}
